File: internal/limiter/fixed_window.go
package limiter

import (
	"context"
	_ "embed"

	"github.com/redis/go-redis/v9"
	redisClient "ratelimiter/internal/redis"
)

// the below line is not a comment it load the lua-script at runtime
// into the below mention variable
//go:embed lua/fixed_window.lua
var fixedWindowScriptSrc string

var fixedWindowScript = redis.NewScript(fixedWindowScriptSrc)

// FixedWindow implements RateLimiter using Redis INCR/EXPIRE over a fixed
// window that resets at the start of every window rather than sliding.
type FixedWindow struct {
	rdb *redisClient.Redis
}

func NewFixedWindow(rdb *redisClient.Redis) *FixedWindow {
	return &FixedWindow{rdb: rdb}
}

func (f *FixedWindow) Allow(ctx context.Context, key string, limit int, window int) (*Result, error) {
	res, err := f.rdb.RunScript(ctx, fixedWindowScript, []string{key}, limit, window)
	if err != nil {
		return nil, err
	}
	vals := res.([]interface{})
	return &Result{
		Allowed:   vals[0].(int64) == 1,
		Remaining: vals[1].(int64),
		ResetAt:   vals[2].(int64),
	}, nil
}

var _ RateLimiter = (*FixedWindow)(nil)

File: internal/limiter/lua/fixed_window.lua
local key = KEYS[1]
local limit = tonumber(ARGV[1])
local window = tonumber(ARGV[2])

-- increment the count for the key
local count = redis.call("INCR", key)

-- if first request, set the expiry window
if count == 1 then
	redis.call("EXPIRE", key, window)
end

local ttl = redis.call("TTL", key)
local now = tonumber(redis.call("TIME")[1])
local reset_at = now + ttl

-- {allowed, remaining, reset_at} | 1 = allowed, 0 = rejected
if count > limit then
	return {0, 0, reset_at}
end
return {1, limit - count, reset_at}
